### PR TITLE
Tracking doc: rewrite `@wdio/selenium-standalone-service` package into TypeScriptTs selenium standlone service

### PR DESCRIPTION
**Pull request by suniljaiswal01**
## Proposed changes
This is a tracking issue for organising the effort of re-writing the code base from vanilla JS into TypeScript. If you are reading this and are interested getting involved, this is a great chance to do so. Here are the rules:

if you are interested converting source files from /packages/wdio-selenium-standalone-service into TypeScript please comment on this thread and mention the files you want to convert to avoid duplicate work
there is no reason to convert the whole package at once - smaller PRs make it easier to review the code
if you make a PR, reference this issue so we can keep track of the progress
if you make a PR, label it with TypeScript and the package name wdio-selenium-standalone-service so we can easier find all PRs
if you make a PR, assign it to the TypeScript Rewrite project

#5844 

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [ ] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] I have added proper type definitions for new commands (if appropriate)

## Further comments

There are still some issue which i need some help in resolving

### CODE REVIEW

File: packages/wdio-selenium-standalone-service/src/launcher.ts [hunk 1]
```diff
@@ -48,11 +48,11 @@ export default class SeleniumStandaloneLauncher {
         this.skipSeleniumInstall = Boolean(options.skipSeleniumInstall)
     }
 
-    async onPrepare(config: Config) :Promise<void> {
+    async onPrepare(config: Config): Promise<void> {
         this.watchMode = Boolean(config.watch)
 
         if (!this.skipSeleniumInstall) {
-            await promisify(SeleniumStandalone.install)(this.installArgs)
+            await promisify(SeleniumStandalone.install.bind(this.installArgs))
```
Comment (suniljaiswal01) [inline packages/wdio-selenium-standalone-service/src/launcher.ts:55]: could someone help me on what should be used here.

Comment (mgrybyk) [reply packages/wdio-selenium-standalone-service/src/launcher.ts:55]: I think it should not be changed

Comment (suniljaiswal01) [reply packages/wdio-selenium-standalone-service/src/launcher.ts:55]: If not changed, then i am getting the below error:

`(property) SeleniumStandaloneLauncher.installArgs: Partial<SeleniumStandalone.InstallOpts> Expected 0 arguments, but got 1.ts(2554)`

Comment (mgrybyk) [reply packages/wdio-selenium-standalone-service/src/launcher.ts:55]: you can define `installSeleniumStandalone = promisify(SeleniumStandalone.install)`
it should have the same type as `SeleniumStandalone.install` but result should be wrapped with `Promise<>`

File: packages/wdio-selenium-standalone-service/src/launcher.ts [hunk 2]
```diff
@@ -1,5 +1,8 @@
 import logger from '@wdio/logger'
 
+declare module 'util' {
+    function promisify<T1, TResult>(fn: (arg1: T1, callback: (err: any, result: TResult) => void) => void): (arg1: T1) => Promise<TResult>;
+}
```
Comment (christian-bromann) [inline packages/wdio-selenium-standalone-service/src/launcher.ts:5]: Why is this necessary?

File: packages/wdio-selenium-standalone-service/src/launcher.ts [hunk 3]
```diff
@@ -16,16 +19,36 @@ const DEFAULT_CONNECTION = {
...
+    outputDir?: string,
+    watch?: boolean,
+}
+
+export interface SeleniumStandaloneOptions {
+    logPath?: string;
+    installArgs?: any;
+    args?: any;
+    skipSeleniumInstall?: boolean;
+}
```
Comment (christian-bromann) [inline packages/wdio-selenium-standalone-service/src/launcher.ts:32]: add new line
```suggestion
}

```

File: packages/wdio-selenium-standalone-service/src/launcher.ts [hunk 4]
```diff
@@ -16,16 +19,36 @@ const DEFAULT_CONNECTION = {
...
 export default class SeleniumStandaloneLauncher {
-    constructor (options, capabilities, config) {
+
```
Comment (christian-bromann) [inline packages/wdio-selenium-standalone-service/src/launcher.ts:34]:
```suggestion
```

File: packages/wdio-selenium-standalone-service/src/launcher.ts [hunk 5]
```diff
@@ -16,16 +19,36 @@ const DEFAULT_CONNECTION = {
...
+    capabilities: any
```
Comment (christian-bromann) [inline packages/wdio-selenium-standalone-service/src/launcher.ts:35]: We should use `Capabilities` from the webdriver package

**Review by christian-bromann — CHANGES_REQUESTED** [review]
Some comments

File: packages/wdio-selenium-standalone-service/src/launcher.ts (hunk 2 above)
Comment (suniljaiswal01) [reply packages/wdio-selenium-standalone-service/src/launcher.ts:5]: There seems to be some issue with promisify overloading , similar issue is mentioned here :
https://github.com/DefinitelyTyped/DefinitelyTyped/issues/20497

By using the above code, we are specifying which method to overload instead of the generic one

**Comment by mgrybyk** [comment]
@suniljaiswal01 please see https://github.com/webdriverio/webdriverio/compare/ts-selenium-standlone-service-fix-types

Based on my comments I've applied some fixes

File: packages/wdio-selenium-standalone-service/package.json [hunk 6]
```diff
@@ -42,6 +41,8 @@
   },
   "types": "selenium-standalone-service.d.ts",
   "devDependencies": {
-    "@types/fs-extra": "^9.0.1"
+    "@types/fs-extra": "^9.0.1",
+    "webdriver": "^6.6.2",
+    "webdriverio": "^6.6.2"
```
Comment (christian-bromann) [inline packages/wdio-selenium-standalone-service/package.json:46]:
```suggestion
    "webdriver": "6.6.2",
    "webdriverio": "6.6.2"
```

File: packages/wdio-selenium-standalone-service/src/@types/webdriverio.d.ts [hunk 7]
```diff
@@ -0,0 +1,5 @@
+declare module WebdriverIO {
+    interface Config {
+        watch?: boolean
```
Comment (christian-bromann) [inline packages/wdio-selenium-standalone-service/src/@types/webdriverio.d.ts:3]: Can we move this to the original typings?

File: packages/wdio-selenium-standalone-service/package.json (hunk 6 above)
Comment (christian-bromann) [reply packages/wdio-selenium-standalone-service/package.json:46]: Also if we only import types from these packages we don't need to declare them as dev dependencies. Given the project setup, TS will find the packages.

File: packages/wdio-selenium-standalone-service/src/@types/webdriverio.d.ts (hunk 7 above)
Comment (suniljaiswal01) [reply packages/wdio-selenium-standalone-service/src/@types/webdriverio.d.ts:3]: @mgrybyk  added this in his commit. Should i remove it?

Comment (christian-bromann) [reply packages/wdio-selenium-standalone-service/src/@types/webdriverio.d.ts:3]: The `watch` property in the config isn't something that is specific to the service, so I would say yes.

**Comment by mgrybyk** [comment]
Don't put watch to Config interface of webdriverio!
We should not allow to set watch option in the config.
@christian-bromann we can have a fix for it in another PR, I'd leave it as is for now.

File: packages/wdio-selenium-standalone-service/package.json (hunk 6 above)
Comment (mgrybyk) [reply packages/wdio-selenium-standalone-service/package.json:46]: Yes! Let's remove both webdriver and webdriverio from here, nice catch!

**Review by mgrybyk — CHANGES_REQUESTED** [review]

File: packages/wdio-selenium-standalone-service/src/@types/webdriverio.d.ts (hunk 7 above)
Comment (mgrybyk) [reply packages/wdio-selenium-standalone-service/src/@types/webdriverio.d.ts:3]: We should think on how to keep it internal only and don't expose to the user.
At least disallow to define it in the config

Comment (mgrybyk) [reply packages/wdio-selenium-standalone-service/src/@types/webdriverio.d.ts:3]: Otherwise it might confuse a user that the option does nothing
https://github.com/webdriverio/webdriverio/blob/e34108325ef1e0d2b8e2306fe407b5d934169777/packages/wdio-config/src/lib/ConfigParser.js#L73

Comment (mgrybyk) [reply packages/wdio-selenium-standalone-service/src/@types/webdriverio.d.ts:3]: We can also define it in webdriverio tpl config like this `readonly watch?: never` and add a comment that it's used for internal purposes only.
@christian-bromann

File: packages/wdio-selenium-standalone-service/.eslintrc.js [hunk 8]
```diff
@@ -0,0 +1,6 @@
+module.exports = {
+    globals: {
+        WebdriverIO: true,
+        WebDriver: true,
+    }
+}
```
Comment (christian-bromann) [inline packages/wdio-selenium-standalone-service/.eslintrc.js:6]: This file can be removed

File: packages/wdio-selenium-standalone-service/src/@types/webdriverio.d.ts (hunk 7 above)
Comment (christian-bromann) [reply packages/wdio-selenium-standalone-service/src/@types/webdriverio.d.ts:3]: sounds good to me

**Comment by christian-bromann** [comment]
if the last 2 outstanding issues are resolved, this will be good to go 👍

**Comment by suniljaiswal01** [comment]
> if the last 2 outstanding issues are resolved, this will be good to go 👍

one is removing the .eslintrc.js file, which is the second? this one:
https://github.com/webdriverio/webdriverio/pull/5913#discussion_r503263687

File: packages/wdio-selenium-standalone-service/.eslintrc.js (hunk 8 above)
Comment (suniljaiswal01) [reply packages/wdio-selenium-standalone-service/.eslintrc.js]: Done.

Comment (suniljaiswal01) [reply packages/wdio-selenium-standalone-service/.eslintrc.js]: can you tell me what does this file do?

**Comment by mgrybyk** [comment]
If you are not getting any errors with eslint file removed - go ahead. Added it because I had some issues with linter

File: packages/wdio-selenium-standalone-service/src/@types/webdriverio.d.ts (hunk 7 above)
Comment (mgrybyk) [reply packages/wdio-selenium-standalone-service/src/@types/webdriverio.d.ts:3]: Second one, I guess

Comment (christian-bromann) [reply packages/wdio-selenium-standalone-service/src/@types/webdriverio.d.ts:3]: Yes, this is an internal flag and only set if you run:

```sh
$ npx wdio wdio.conf.js --watch
```

as the `watch` flag will be merged into the config.

**Comment by christian-bromann** [comment]
@suniljaiswal01 correct, instead of defining custom WebdriverIO types, add the `watch` property to the WebdriverIO.Config definition.

**Comment by suniljaiswal01** [comment]
Thank you for the help @christian-bromann @mgrybyk . Let me know if anything else need to be change

File: scripts/templates/webdriverio.tpl.d.ts [hunk 9]
```diff
@@ -633,7 +633,9 @@ declare namespace WebdriverIO {
         // ... browser commands ...
     }
 
-    interface Config extends Options, Omit<WebDriver.Options, "capabilities">, Hooks {}
+    interface Config extends Options, Omit<WebDriver.Options, "capabilities">, Hooks {
+        watch?: boolean
```
Comment (mgrybyk) [inline scripts/templates/webdriverio.tpl.d.ts:637]:
```suggestion
         /**
         * internal usage only. To run in watch mode see https://webdriver.io/docs/watcher.html
         */
        watch?: never;
```

**Review by mgrybyk — CHANGES_REQUESTED** [review]
Got one minor comment, other than that LGTM.

Thanks a lot for your hard work and patience!

File: scripts/templates/webdriverio.tpl.d.ts (hunk 9 above)
Comment (mgrybyk) [reply scripts/templates/webdriverio.tpl.d.ts:637]: User should not be able to set it in `wdio.conf.js`
image

Comment (suniljaiswal01) [reply scripts/templates/webdriverio.tpl.d.ts]: using never type for watch property is failing the execution, we are using `await launcher.onPrepare({ watch: true })` in test file

Comment (suniljaiswal01) [reply scripts/templates/webdriverio.tpl.d.ts]: changing `await launcher.onPrepare({ watch: true })`  to `await launcher.onPrepare({ watch: true } as never)` is fixing the failing test. is this ok?

Comment (mgrybyk) [reply scripts/templates/webdriverio.tpl.d.ts]: Fine for me

**Review by christian-bromann — APPROVED** [review]
👍 thank you!

@mgrybyk any outstanding issues?

**Review by mgrybyk — APPROVED** [review]

**Comment by christian-bromann** [comment]
Great job @suniljaiswal01 !
